runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -1,14 +1,269 @@
+                    } catch (Exception e) {
+                        log.error("Error while attempting to get a valid subscription timestamp.  Retrying.", e);
+                        throw new RetryNeededException();
+                    }
+                }).setOptions(x -> {
+                    x.setBackoffDuration(Duration.ofMillis(backoff_retry_duration_ms));
+                    x.setMaxRetryThreshold(Duration.ofMinutes(max_duration_threshold_mins));
+                }).run();
+            } catch (InterruptedException e) {
+                log.error("Failed to get a valid subscription timestamp.", e);
+                clientListener.onError(e);
+            }
+            return null;
+        }
+
+        /**
+         * Gets the latest token from the sequencer and builds a timestamp from it
+         * @return Timestamp
+         */
+        private CorfuStoreMetadata.Timestamp getTimestamp() {
+            Token token = corfuStore.getRuntime().getSequencerView().query().getToken();
+            return CorfuStoreMetadata.Timestamp.newBuilder()
+                .setEpoch(token.getEpoch())
+                .setSequence(token.getSequence())
+                .build();
+        }
+
+        /**
+         * Checks if snapshot sync is ongoing at the given timestamp.
+         * @param timestamp
+         * @return
+         */
+        private boolean checkSnapshotSyncOngoing(CorfuStoreMetadata.Timestamp timestamp) {
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE, IsolationLevel.snapshot(timestamp))) {
+                List<CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message>> entries =
+                    txn.executeQuery(REPLICATION_STATUS_TABLE, p -> true);
+                // In LR V1, it is a valid assumption that the size of replication status table will be 1 as there is
+                // only 1 remote cluster.  This implementation will change in LR V2
+                Preconditions.checkState(entries.size() == 1);
+
+                CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message> entry = entries.get(0);
+                txn.commit();
+                return entry.getPayload().getDataConsistent() ? false : true;
+            } catch (TransactionAbortedException e) {
+                // Since this is a read-only transaction, the abort can only be if this version of the table is not
+                // available in the JVM's cache.  The underlying cause for this error is TrimmedException
+                Preconditions.checkState(e.getCause() instanceof TrimmedException);
+
+                // This means that there have been updates to the Replication Status table after the above timestamp,
+                // and this snapshot has been evicted from the cache of the JVM.  We cannot tell the status of snapshot
+                // sync at this timestamp.  Return true so that caller retries the check with a later timestamp.
+                return true;
```

Comment:
We are actually not retrying immediately. We start listening for replication_status update, looking for isDataConsistent.

We can run into the case where we come here at time T1 because we are on old version that has been evicted from cache, and the version we have right now is T2. But snapshot sync was completed at time T0.
We then keep waiting for another snapshot_sync to get triggered and completed, which may or may not happen.

Reply:
maybe return another detail that says the timestamp needs to be refreshed and then retried?

Reply:
good point.  We need to distinguish between the case where the state at T1 was not available as it was evicted from the cache vs snapshot sync was indeed in progress.  

Changed this to throw the exception instead of returning true.  The exception will get caught in the retry block and a retry will be triggered.